### ISSUE TITLE
local ToUpper mapping under "use utf8"

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#3446]
Migrated from [rt.perl.org#5860](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=5860) (status was 'resolved')

Searchable as RT5860$

**Comment by p5pRT** [comment]
### From alberto@head-cfa.harvard.edu



I am trying to override the built\-in ToUpper functions provided in
utf8\_heavy\.pl \(because I want to define my own case mapping\)\, and
things don't seem to work properly\.  Here is a test little program
that is supposed to only upcase the "a" letters&#8203;:

use Carp;
use Devel&#8203;::Peek;

sub ToUpper \{
&nbsp;    carp "called function "\, \(caller\(0\)\)\[3\];

&nbsp;    return \<\<'END';
0061        0041
END
\}

while \(\<>\) \{
&nbsp;    use utf8;
&nbsp;    chop;
&nbsp;    print $\_\, " \-> ";
&nbsp;    s/\(\\w\+\)/\\U$1/g;
&nbsp;    print $\_\, "\\t\("\, uc\($\_\)\, "\)\\n";
\}


this is what happens when I run it&#8203;:

alberto@&#8203;guinan\-584&#8203;: bin/perl BUGS/utf8\-case\.pl 
this and that
this and that \-> called function main&#8203;::ToUpper at /proj/ads/soft/utils/src/perl\-current/lib/5\.6\.1/utf8\_heavy\.pl line 30
THIS 006 THAT   \(THIS 006 THAT\)
that and the other
that and the other \-> thAt And the other        \(thAt And the other\)


So the first time my ToUpper function is called to create the
uppercase mappings\, the input string "this and that" is incorrectly
mapped to "THIS 006 THAT"\.  The second string I type in \("that and the
other"\) is correctly mapped\.  This seems to happen no matter what the
first and second strings are\.

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.1:

Configured by alberto at Mon Feb 19 11:57:20 EST 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=solaris, osvers=2.6, archname=sun4-solaris
    uname='sunos guinan 5.6 generic_105181-11 sun4u sparc '
    config_args='-des -Dusedevel=true -Dcc=gcc -Dprefix=/proj/ads/soft/utils/src/perl-current -Dlocincpth=/proj/ads/soft/utils/include /opt/local/include -Dloclibpth=/proj/ads/soft/utils/lib /opt/local/lib -Uinstallusrbinperl'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='gcc', ccflags ='-I/proj/ads/soft/utils/include -I/opt/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O',
    cppflags='-I/proj/ads/soft/utils/include -I/opt/local/include'
    ccversion='', gccversion='2.8.1', gccosandvers='solaris2.5.1'
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=4321
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=16
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=8, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='gcc', ldflags =' -L/proj/ads/soft/utils/lib -L/opt/local/lib '
    libpth=/proj/ads/soft/utils/lib /opt/local/lib /usr/lib /usr/ccs/lib
    libs=-lsocket -lnsl -lgdbm -ldb -ldl -lm -lc
    perllibs=-lsocket -lnsl -ldl -lm -lc
    libc=/lib/libc.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags=' '
    cccdlflags='-fPIC', lddlflags='-G -L/proj/ads/soft/utils/lib -L/opt/local/lib'

Locally applied patches:
    v5.6.1-TRIAL2


@INC for perl v5.6.1:
    /proj/ads/soft/utils/src/perl-current/lib/5.6.1/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/5.6.1
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.6.1/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.6.1
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.6.0/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.6.0
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00556.hold/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00556.hold
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00555.hold/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00555.hold
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00554.hold/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00554.hold
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00553.hold/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00553.hold
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00552.hold/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00552.hold
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.00551.hold
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.005/sun4-solaris
    /proj/ads/soft/utils/src/perl-current/lib/site_perl/5.005
    /proj/ads/soft/utils/src/perl-current/lib/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home/alberto
    LANG=C
    LANGUAGE (unset)
    LD_LIBRARY_PATH=/proj/ads/soft/utils/lib:/opt/SUNWspro/lib:/opt/X11R5/lib:/opt/local/lib/mh
    LOGDIR (unset)
    PATH=/home/tex/teTeX/bin/sparc-solaris:/proj/ads/soft/utils/bin:/proj/ads/soft/articles/bin:/usr/sbin:/opt/local/bin:/opt/NeWSprint/bin:/opt/local/bin/mh:/usr/bin:/usr/openwin/bin:/opt/X11R5/bin:/usr/ccs/bin:/usr/ucb:/usr/local/bin:/usr/afsws/bin:/home/alberto/bin:.:/soft/tex/bin:/proj/ads/soft/articles/bin/solaris:/data/guinan/soft/bin
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @jhi

On Mon\, Feb 19\, 2001 at 05&#8203;:56&#8203;:23PM \-0500\, alberto@&#8203;head\-cfa\.harvard\.edu wrote&#8203;:
> This is a bug report for perl from alberto@&#8203;guinan\.harvard\.edu\,
> generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.
> 
> 
> \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> \[Please enter your report here\]
> 
> I am trying to override the built\-in ToUpper functions provided in
> utf8\_heavy\.pl \(because I want to define my own case mapping\)\, and
> things don't seem to work properly\.  Here is a test little program
> that is supposed to only upcase the "a" letters&#8203;:

I agree that your example exhibits erroneous behaviour \-\- but I'm a
little bit lost as to why did you expect it to work?  In other words\,
is defining one's own ToUpper documented somewhere?  I agree that
being able to define such things would be nice\, but I wouldn't endorse
spending too much time with the deep guts of utf8\*\{\.pm\,pl\}\, that is
all undocumented \(I hope\!\) for a bunch of reasons\.  As far as I know
at the moment there is no Officially Supported Way of redefining the
meanings of \\U et alia\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

In message \<20010222080959\.K21295@&#8203;chaos\.wustl\.edu>\, Jarkko Hietaniemi writes&#8203;:

> On Mon\, Feb 19\, 2001 at 05&#8203;:56&#8203;:23PM \-0500\, alberto@&#8203;head\-cfa\.harvard\.edu wrote&#8203;:
> > This is a bug report for perl from alberto@&#8203;guinan\.harvard\.edu\,
> > generated with the help of perlbug 1\.33 running under perl v5\.6\.1\.
> > 
> > 
> > \-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-
> > \[Please enter your report here\]
> > 
> > I am trying to override the built\-in ToUpper functions provided in
> > utf8\_heavy\.pl \(because I want to define my own case mapping\)\, and
> > things don't seem to work properly\.  Here is a test little program
> > that is supposed to only upcase the "a" letters&#8203;:
> 
> I agree that your example exhibits erroneous behaviour \-\- but I'm a
> little bit lost as to why did you expect it to work?  In other words\,
> is defining one's own ToUpper documented somewhere?  I agree that
> being able to define such things would be nice\, but I wouldn't endorse
> spending too much time with the deep guts of utf8\*\{\.pm\,pl\}\, that is
> all undocumented \(I hope\!\) for a bunch of reasons\.  As far as I know
> at the moment there is no Officially Supported Way of redefining the
> meanings of \\U et alia\.

Jarkko\,

I agree that I was going down the "unsupported feature" path here in
redefining ToUpper and ToLower\.  I have to say\, though\, that the latest
Camel book does invite people to do things such as defining one's own
Unicode character properties \(quite handy thing\!\) and points you to the
directory PATH\_TO\_PERLLIB/unicode/Is\.  Once you get there\, it's hard
not to look around and notice the existence of the To directory\, which
is where I saw the definition of ToUpper\.

Anyway\, let me just tell you why I was doing this\, which may explain
the reason why somebody may want to fix the problem&#8203;: brute\-force
uppercasing of Unicode characters more often than not gives you
unpleasant results \(as pointed out by Tim Bray\)\.  Just picture what
happens when you have a scientific paper and you map all lowercase
greek letters to uppercase&#8203;: most of your math formulas would stop
making any sense\.  So in the context of full\-text indexing \(which is
what I'm working on\)\, I want to have fine control on what characters
get uppercased\.  I have been doing this by explicitely specifying
character ranges with s/// \(tr was still broken under 5\.6\.0 last time
I tried it\)\, but thought the ToUpper approach would be quite elegant
and much faster\, and one that if available would be touted by PERL
users as the ultimate hack\.

Thanks for your help\,


\-\- Alberto


\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
Alberto Accomazzi                          mailto&#8203;:aaccomazzi@&#8203;cfa\.harvard\.edu
NASA Astrophysics Data System                      http&#8203;://adsabs\.harvard\.edu
Harvard\-Smithsonian Center for Astrophysics        http&#8203;://cfawww\.harvard\.edu
60 Garden Street\, MS 83\, Cambridge\, MA 02138 USA   
\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*\*
